pykafka/producer.py

```diff
@@ -237,7 +428,167 @@ def produce(self, messages):
+        :type linger_ms: int
+        :param acquire: Whether to acquire and release the lock. If False,
+            this function assumes that the lock is already held by the caller
+            and raises an AssertionError if not.
+        :type acquire: bool
+        :param release_inflight: Whether to decrement the messages_inflight
+            counter when the queue is flushed. True means that the messages
+            popped from the queue will be discarded unless re-enqueued
+            by the caller.
+        :type release_inflight: bool
+        """
+        assert acquire or self.lock.locked()
+        self._wait_for_flush_ready(linger_ms, acquire=acquire)
+        if acquire:
+            self.lock.acquire()
+        batch = [self.queue.pop() for _ in xrange(len(self.queue))]
+        if release_inflight:
+            self.messages_inflight -= len(batch)
+        if not self.slot_available.is_set():
+            self.slot_available.set()
+        if acquire:
+            self.lock.release()
+        return batch
+
+    def _wait_for_flush_ready(self, linger_ms, acquire=True):
+        """Block until the queue is ready to be flushed
+
+        If the queue does not contain at least one message after blocking for
+        `linger_ms` milliseconds, return.
+
+        :param linger_ms: How long (in milliseconds) to wait for the queue
+            to contain messages before returning
+        :type linger_ms: int
+        :param acquire: Whether to acquire and release the lock. If False,
+            this function assumes that the lock is already held by the caller
+            and raises an AssertionError if not.
+        :type acquire: bool
+        """
+        assert acquire or self.lock.locked()
+        if len(self.queue) == 0:
+            if acquire:
+                self.lock.acquire()
+            if len(self.queue) == 0:
+                self.flush_ready.clear()
+            if acquire:
+                self.lock.release()
+            self.flush_ready.wait(linger_ms / 1000)
+
+    def _wait_for_slot_available(self, acquire=True):
+        """Block until the queue has at least one slot not containing a message
+
+        :param acquire: Whether to acquire and release the lock. If False,
+            this function assumes that the lock is already held by the caller
+            and raises an AssertionError if not.
+        :type acquire: bool
+        """
+        assert acquire or self.lock.locked()
+        if len(self.queue) >= self.producer._max_queued_messages:
+            if acquire:
+                self.lock.acquire()
+            if len(self.queue) >= self.producer._max_queued_messages:
+                self.slot_available.clear()
+            if acquire:
+                self.lock.release()
+            if self.producer._block_on_queue_full:
+                self.slot_available.wait()
+            else:
+                raise ProducerQueueFullError("Queue full for broker %d",
```

Comment:
When I call `produce` with a bunch of messages and this happens, is there an easy way to find out which of my messages got queued to an OwnedBroker and which didn't?